Crash site rocket tanks

Jayefuu opened the pull request:
Adds research, unlockable from the market, that allows tanks to fire rockets at enemy worms and spawners.

The intention is to make mid-game more fun and make tanks more useful, they're often ignored while players wait for spidertrons. Intentionally didn't allow rockets to target biters and spitters because the part of the game I want them to be enticing for is when we have few rockets just after we get the science to make them.

They don't target enemy turrets because I don't want to change the destroyer spawning behaviour.

The first level of tech allows rockets to be shot at a 3 second interval. Upgradable to level 4, 0.5 seconds per rocket.

Almost ready for a live test, one small feature to complete.

Jayefuu committed:
Rocket tanks first  draft

Jayefuu commented on map_gen/maps/crash_site/rocket_tanks.lua:
```diff
@@ -0,0 +1,153 @@
+    'big-worm-turret',
+    'behemoth-worm-turret'
+}
+--[[local function on_tick()
```
Old function I used before using the removable nth tick event.

Jayefuu commented on map_gen/maps/crash_site/rocket_tanks.lua:
```diff
@@ -0,0 +1,153 @@
+end
+)
+
+local rocket_tank_level_intervals = {
```
Used a table rather than maths to determine the firing interval because I wanted some sub-one second options.

Jayefuu commented on map_gen/maps/crash_site/rocket_tanks.lua:
```diff
@@ -0,0 +1,153 @@
+    end
+    if interval_level == 5 then
+        -- This doesn't work....
+        Retailer.remove_item('Spawn', 'rocket_tanks_fire_rate')
```
@grilledham or @SimonFlapse I need your help with Retailer.remove_item, I couldn't get it to work from the documentation and examples I found. Any help much appreciated. I think this is the last thing I need to change before I can do a live play test.

grilledham replied:
```suggestion
        Retailer.remove_item(market_id, 'rocket_tanks_fire_rate')
```

grilledham commented on map_gen/maps/crash_site/rocket_tanks.lua:
```diff
@@ -0,0 +1,153 @@
+local on_tick
+on_tick = Token.register(
+    function()
+    for k, v in pairs(tank_entities) do
+        local unit_number = k
+        local tank = v
```
```suggestion
    for unit_number, tank in pairs(tank_entities) do
```

grilledham replied:
I'd add a check that tank is valid, that way this doesn't depend on the order in which events are called.

grilledham commented on map_gen/maps/crash_site/scenario.lua:
```diff
@@ -787,6 +791,14 @@ local function init(config)
+                name = 'rocket_tanks_fire_rate',
+                name_label = {'command_description.crash_site_rocket_tanks_name_label', 1},
+                sprite = 'item-group/production',
+                description = 'Upgrade the tank rocket firing rate to reduce the rocket interval.\n\nPlace rockets in the tank inventory to have them automatically target enemy worms and nests.'
```
Use localised string

grilledham reviewed (commented):
I've had a quick look, I'll play test later.

Jayefuu committed:
Final updates

- Made changes suggested by grilledham
- Disabled market item when research levels complete so that the tooltip can still be seen
- Changed cost to 7500 extra per level

grilledham committed:
Updates:
- update .luacheckrc.
- Add stack limits to airstrike and rocket tank market items.
- Refactor rocket_tanks.lua

grilledham commented:
I made some changes, I was going to suggest them as part of a review, but I needed to make changes to files not in the review for it to work.

grilledham committed:
Fix lua check

grilledham committed:
Fix broken item.name check

Jayefuu committed:
Updated rocket tank balance

- Reduced intervals for each level from 3, 2, 1, 0.5 seconds to 2, 1.5, 1, 0.5 seconds.
- Reduced cost from 7500 per level to 4000 per level.

These changes should make the tank useful in the period of the game for which it's designed. I will watch the next few games to check balance and adjust if it's too OP.
